pkg/rbd/rbd.go

```diff
@@ -27,10 +27,89 @@ import (
 	"k8s.io/utils/exec"
 )
 
-// PluginFolder defines the location of rbdplugin
+/*
+RADOS omaps usage:
+
+This note details how we preserve idempotent nature of create requests and retain the relationship
+between orchestrator (CO) generated Names and plugin generated names for images and snapshots
+
+The implementation uses Ceph RADOS omaps to preserve the relationship between request name and
+generated image (or snapshot) name. There are 4 types of omaps in use,
+- A "csi.volumes.directory.[csi-id]" (or volOMap+.+CsiInstanceID), we call this the volOMap
+  - stores keys named using the CO generated names for volume requests
+  - keys are named "csi.volume."+[CO generated VolName]
+  - Key value contains the RBD image name that is created or will be created, for the CO provided name
+
+- A "csi.snaps.directory.[csi-id]" (or snapsOMap+.+CsiInstanceID), we refer to this as the snapsOMap
+  - stores keys named using the CO generated names for snapshot requests
+  - keys are named "csi.snap."+[CO generated SnapName]
+  - Key value contains the RBD snapshot name that is created or will be created, for the CO provided name
+
+- A per image omap named "csi.volume."+[RBD image name], we refer to this as the imgOMap
+  - stores a single key named "csi.volname", that has the value of the CO generated VolName that
```

Comment:
The ```ListSnapshotsRequest``` CSI method has an optional ```SourceVolumeId``` field [1]. I assume that this would need to be a volume ID from this CSI, so should this also include a collection of snapshot ids? When combined w/ the comment below about ensuring that this object doesn't already exist, you can keep this object in the pool until both the volume and all associated snapshots have been removed if that ```SourceVolumeId``` link should remain valid even after the volume is removed.

[1] https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L3215

Reply:
Yes, the assumption is correct, `ListSnapshotRequest` can come with a `SourceVolumeID` that is a request to only list snapshots of that source volume.

I was going to list all snapshots of the provided `SourveVolumeID` and filter out those created by the CSI plugin, and return the same in such a request (when implementing `ListSnapshots` capability i.e).

The above manner of listing assumes that the RBD image that backs the SourceVolumeID exists, when such a request comes. To overcome the possibility that we will support source image deletion if it has snapshots, the snapshot names need to be preserved in the `csi.volume.<image-uuid>` omap as suggested.

Currently though we do not allow delete of an image if it has snapshots, and way forward is being discussed in other issues and PRs.

What I propose for this PR is,
- Postpone stashing snapshot names in the volume omap and the process of retaining the omap until the list of keys within it is empty, till we decide on a course for the delete
OR,
- Just to ensure data is already stored on disk and prevent future incompatibility, add the snapshot name into the `csi.volume.<image-uuid>` omap, even if this is unused, but do not deal with retaining the omap on delete of the source image, which can be plumbed in later once the delete of source image with snapshots is considered.

Thoughts?

Reply:
I guess we can avoid storing it for now and when support is added for deleting volumes w/ snapshots, the necessary lookups can be written to RADOS when the volume deletion request is processed.

Reply:
Updated [comment](https://github.com/ceph/ceph-csi/pull/288#issuecomment-476837855) tracking TODO and TechDebt with List support line item, including a note on using the image omap to store the snapshot names.